src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp

```diff
@@ -61,10 +65,126 @@ double overlap_width(const size_t overlap_extent,
+  return volume_offset_;
+}
+
+size_t OverlapIterator::overlap_offset() const noexcept {
+  return overlap_offset_;
+}
+
+void OverlapIterator::reset() noexcept {
+  volume_offset_ = initial_offset_;
+  overlap_offset_ = 0;
+  stride_count_ = 0;
+  slice_ = 0;
+}
+
+namespace detail {
+
+template <size_t Dim>
+void data_on_overlap_impl(double* overlap_data, const double* volume_data,
+                          const size_t num_components,
+                          const Index<Dim>& volume_extents,
+                          const size_t overlap_extent,
+                          const Direction<Dim>& direction) noexcept {
+  const size_t volume_num_points = volume_extents.product();
+  const size_t overlap_num_points = LinearSolver::Schwarz::overlap_num_points(
+      volume_extents, overlap_extent, direction.dimension());
+  for (OverlapIterator overlap_iterator{volume_extents, overlap_extent,
+                                        direction};
+       overlap_iterator; ++overlap_iterator) {
+    for (size_t j = 0; j < num_components; ++j) {
+      const size_t volume_data_index =
+          overlap_iterator.volume_offset() + j * volume_num_points;
+      const size_t overlap_data_index =
+          overlap_iterator.overlap_offset() + j * overlap_num_points;
+      // clang-tidy: do not use pointer arithmetic
+      // NOLINTNEXTLINE
+      overlap_data[overlap_data_index] = volume_data[volume_data_index];
+    }
+  }
+}
+
+template <size_t Dim>
+void add_overlap_data_impl(double* volume_data, const double* overlap_data,
+                           const size_t num_components,
+                           const Index<Dim>& volume_extents,
+                           const size_t overlap_extent,
+                           const Direction<Dim>& direction) noexcept {
+  const size_t volume_num_points = volume_extents.product();
+  const size_t overlap_num_points = LinearSolver::Schwarz::overlap_num_points(
+      volume_extents, overlap_extent, direction.dimension());
+  for (OverlapIterator overlap_iterator{volume_extents, overlap_extent,
+                                        direction};
+       overlap_iterator; ++overlap_iterator) {
+    for (size_t j = 0; j < num_components; ++j) {
+      const size_t volume_data_index =
+          overlap_iterator.volume_offset() + j * volume_num_points;
+      const size_t overlap_data_index =
+          overlap_iterator.overlap_offset() + j * overlap_num_points;
+      // clang-tidy: do not use pointer arithmetic
+      // NOLINTNEXTLINE
+      volume_data[volume_data_index] += overlap_data[overlap_data_index];
+    }
+  }
+}
+
+}  // namespace detail
+
 /// \cond
 #define DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
-#define INSTANTIATE(r, data) \
-  template size_t overlap_num_points(const Index<DIM(data)>&, size_t, size_t);
+#define INSTANTIATE(r, data)                                                   \
+  template size_t overlap_num_points(const Index<DIM(data)>&, size_t, size_t); \
```

Comment:
I know it's not from this PR, but since you're touching this add the `noexcept` to this instantiation.